⚠️ helm: use chunking release driver in systemNamespace

joelanford opened the pull request:
<!--
Please prefix the title of this PR with one of the following icons:

    * ⚠ (:warning:, major/breaking change)
    * ✨ (:sparkles:, minor/compatible change)
    * 🐛 (:bug:, patch/bug fix)
    * 📖 (:book:, docs)
    * 🌱 (:seedling:, other)

-->

# Description

This PR:
1. uses a new helm release storage driver from `helm-operator-plugins` that chunks release data into one or more secrets (this also incidentally means that our client and the Helm CLI client won't mistake each other's release storage for their own).
2. moves release secret storage back to our system namespace so that we don't leak details of our release bookkeeping to users.
3. Notes that the new helm-operator-plugins commit also makes it possible to use one service account for release storage and a separate service account for deploying bundle contents. This is necessary if we want to avoid leaking our internal release storage mechanism to users (who would otherwise have to configure RBAC for release storage permission)


Fixes: https://github.com/operator-framework/operator-controller/issues/923
<!--
Thank you for your contribution!

Please provide a summary of the changes and the motivation behind the change.
-->

## Reviewer Checklist

- [ ] API Go Documentation
- [ ] Tests: Unit Tests (and E2E Tests, if appropriate)
- [ ] Comprehensive Commit Messages
- [ ] Links to related GitHub Issue(s)

joelanford commented on internal/httputil/httputil.go:
```diff
@@ -7,7 +7,7 @@ import (
 
 func BuildHTTPClient(cpw *CertPoolWatcher) (*http.Client, error) {
-	httpClient := &http.Client{Timeout: 10 * time.Second}
+	httpClient := &http.Client{Timeout: 2 * time.Minute}
```
Now that we have the operatorhubio catalog installed by default, our e2e tests are entering and staying in exponential backoff because it takes longer for the operatorhubio catalog to be pulled, unpacked, and cached than the test catalog does.

Once we have #1028, we could consider excluding the operatorhubio catalog from resolution during these e2es. Even so, we likely want this value bumped up a bit to handle the default catalog without requests timing out.

tmshort replied:
Can you add a TODO to revert this later?

joelanford replied:
I reverted this. Seems to be a problem locally for me only.

tmshort commented on go.mod:
```diff
@@ -17,7 +17,7 @@ require (
 	github.com/operator-framework/api v0.26.0
 	github.com/operator-framework/catalogd v0.20.0
-	github.com/operator-framework/helm-operator-plugins v0.3.1
+	github.com/operator-framework/helm-operator-plugins v0.3.1-0.20240722173643-2e18c5b70f9c
```
release?

joelanford replied:
Yeah, talking about it over on Kubernetes Slack, here: https://kubernetes.slack.com/archives/C0181L6JYQ2/p1722268675418209?thread_ts=1722268591.413059&cid=C0181L6JYQ2

acornett21 replied:
You're now unblocked here, see below
https://github.com/operator-framework/helm-operator-plugins/releases/tag/v0.4.0

tmshort commented on cmd/manager/main.go:
```diff
@@ -161,40 +159,21 @@ func main() {
-		helmclient.RestConfigMapper(restConfigMapper),
+		helmclient.StorageDriverMapper(action.ChunkedStorageDriverMapper(coreClient, mgr.GetAPIReader(), systemNamespace)),
+		helmclient.ClientNamespaceMapper(func(obj client.Object) (string, error) {
+			ext := obj.(*ocv1alpha1.ClusterExtension)
```
could this fail? do you need an `ok` on this type conversion?

tmshort replied:
(Yes, I see it's identical to the prior code that was defined previously)

joelanford replied:
If this fails, we did something horribly wrong (i.e. we used the mapper function in a reconciler for a different type), so a panic is reasonable, IMO. There is nothing a user could do to fix it.

acornett21 commented on config/base/rbac/role_binding.yaml:
```diff
@@ -10,3 +10,17 @@ subjects:
+subjects:
+  - kind: ServiceAccount
+    name: controller-manager
+    namespace: system
```
Nit: No new line at the end of this file.

joelanford replied:
Fixed here: https://github.com/operator-framework/operator-controller/commit/bebe28ccf7b4808e046401a36b8aa6af34fdd415

tmshort reviewed (approved):
/lgtm
Looks reasonable to me
(Added a comment later about a TODO though)

tmshort reviewed (changes requested):
OK, I want the TODO in there about the timeout.
I would prefer a release of helm-operator-plugins, failing that, perhaps a TODO.

tmshort commented:
And... a rebase is needed...

joelanford commented:
Ok, rebased and reverted the catalog client timeout change, so it is back to 10s. Looks like e2e still passes in GH CI.

I'm also seeing `main` fail the same as locally for me as this branch, so let's chalk this up as a "me" problem for now.

tmshort reviewed (approved):
/lgtm

tmshort reviewed (approved):
/lgtm
